okapi/services/draftlogs/upload_fieldnotes_file/WebService.php

```diff
@@ -0,0 +1,175 @@
+                $user_id     = $request->token->user_id;
+                $geocache_id = $geocache['internal_id'];
+                $type        = Okapi::logtypename2id($n['type']);
+                $date        = date("Y-m-d H:i:s", strtotime($n['date']));
+                $text        = $n['log'];
+
+                Db::query("
+                    insert into field_note (
+                        user_id, geocache_id, type, date, text
+                    ) values (
+                        '".Db::escape_string($user_id)."',
+                        '".Db::escape_string($geocache_id)."',
+                        '".Db::escape_string($type)."',
+                        '".Db::escape_string($date)."',
+                        '".Db::escape_string($text)."'
+                    )
+                ");
+
+            }
+            $result = array(
+                'success'          => true,
+                'totalRecords'     => $notes['totalRecords'],
+                'processedRecords' => $notes['processedRecords']
+            );
+            //$result = json_encode($notes, JSON_PRETTY_PRINT); // debug
+        }
+        return Okapi::formatted_response($request, $result);
+    }
+
+    // ------------------------------------------------------------------
+
+    private static function parse_notes($field_notes)
+    {
+        $decoded_field_notes = base64_decode($field_notes, true);
+        if ($decoded_field_notes === false) return false;
+
+        $multiline = self::fieldNotesTxtArea2Array($decoded_field_notes);
+        $submittable_logtype_names = Okapi::get_submittable_logtype_names();
+        $records          = [];
+        $totalRecords     = 0;
+        $processedRecords = 0;
+
+        foreach ($multiline as $line) {
+            $totalRecords++;
+            $line = trim($line);
+            $fields = self::CSVtoArray($line);
+
+            $code = $fields[0];
+            $date = $fields[1];
+            $type = $fields[2];
+
+            if (!in_array($type, $submittable_logtype_names)) continue;
+
+            $log  = nl2br($fields[3]);
+
+            $records[] = [
+                'code' => $code,
+                'date' => $date,
+                'type' => $type,
+                'log'  => $log,
+            ];
+            $processedRecords++;
+        }
+        return ['success' => true, 'records' => $records, 'totalRecords' => $totalRecords, 'processedRecords' => $processedRecords];
+    }
+
+
+    // ------------------------------------------------------------------
+
+    private static function fieldNotesTxtArea2Array($fieldnotes)
+    {
+        $output = [];
+        $buffer = '';
+        $start = true;
+
+        $lines = explode("\n", $fieldnotes);
+        $lines = array_filter($lines); // Drop empty lines
+
+        foreach ($lines as $line) {
+            if ($start) {
+                $buffer = $line;
+                $start = false;
+            } else {
+                if (strpos($line, 'OC') !== 0) {
```

Comment:
I don't understand how newlines are handled in this file. How can you recognize "newline which is part of the log entry" from "newline which separates two CSV rows"?

```
OC1234,Date,Found it,This first line of multiline log.
OC is awesome! That's the second line of a multiline log.
```

Reply:
See my detailed explanation above regarding the format and rationales for using my own custom functions.

BTW: I also spent hours passing the file, which is actually a long string, directly as utf-8 string in the POST body. I have not accomplished to generate a Oauth Level 3 signature ever. Once I base64 encoded it, the problems went away instantly. I understand that all you had back then was Oauth1 but I understand even much much more why Oauth2 was invented.

Reply:
This seems to be a problem with your OAuth signing library then? There are many other methods which take UTF-8 strings and they do not require Base64.

BTW, as you mentioned before, you can use `PLAINTEXT` authentication with https, so you do not need to sign your requests anymore.

Reply:
> This seems to be a problem with your OAuth signing library then? There are many other methods which take UTF-8 strings and they do not require Base64.

I am not using any library, because I couldn't get any to work in node.js. If you're really curious I can send you my testing code, it is just a few lines of JS then you can perhaps see that my signature calculation is correct and still it dies not work if the non base64 encoded string is in there.
 
> BTW, as you mentioned before, you can use `PLAINTEXT` authentication with https, so you do not need to sign your requests anymore.

The OCDE docker based dev env runs only over http. PLAINTEXT is not an option for my testing, unfortunately.

BTW: http or https, PLAINTEXT or HMAC: If I try to put a multiline string into the OKAPI Brower's entry field, how is that supposed to work? No problem with a base64 string.

Reply:
You can hack it locally, so that it thinks that it's https. See `OkapiOAuthServer.php` line 16.

I think OKAPI Browser doesn't support multiple, sadly. I won't be able to release a new version.

Reply:
> You can hack it locally, so that it thinks that it's https. See `OkapiOAuthServer.php` line 16.

Good pointer! Thx much for your help, not just here, but in general. I treally appreciate it a lot.

First I will revisit the CSV parsing on the server side and once that really works without my custom function I will try to send PLAINTEXT signed non-base64 data and see how it goes.

> 
> I think OKAPI Browser doesn't support multiple, sadly. I won't be able to release a new version.

No problem. It is of great help the way it is. For development I use the base64. Should we figure out at the end, how to do without it, it is a two line change to remove the base64 decoding step.

Reply:
Happy to help!